src/define_component.ts

```diff
@@ -0,0 +1,99 @@
+import { Wunderbaum } from "./wunderbaum";
+
+function appendCssLink(url: string, shadowRoot?: ShadowRoot) {
+  let link = document.createElement("link");
+  link.setAttribute("rel", "stylesheet");
+  link.setAttribute("href", url);
+  if (shadowRoot) {
+    shadowRoot.appendChild(link);
+  } else {
+    document.head.appendChild(link);
+  }
+}
+
+class WunderbaumComponent extends HTMLElement {
+  // class WunderbaumTree extends HTMLDivElement {
+  protected _tree?: Wunderbaum;
+  protected _treeElement: HTMLDivElement;
+
+  // fires when an instance of the element is created
+  constructor() {
+    // NOTE: We would rather derrive from HTMLDivElement instead of HTMLElement, like here:
+    //   https://html.spec.whatwg.org/multipage/custom-elements.html#custom-elements-customized-builtin-example
+    // but that does not seem to work in Safari (`super()` fails).
+    // So we use HTMLElement and add a div as child element.
+
+    // establish prototype chain
+    super();
+
+    // attaches shadow tree and returns shadow root reference
+    // https://developer.mozilla.org/en-US/docs/Web/API/Element/attachShadow
+    const shadowRoot = this.attachShadow({ mode: "open" });
+
+    const url = new URL(import.meta.url);
+    let parts = url.pathname.split("/");
+    parts[parts.length - 1] = "wunderbaum.css";
+    const path = parts.join("/");
+    console.log(`Wunderbaum component using CSS path: ${path}`);
+
+    if (this.hasAttribute("icon-css")) {
+      appendCssLink(this.getAttribute("icon-css")!, shadowRoot);
+    }
+    appendCssLink(path, shadowRoot);
+
+    // @import url("${path}");
+    shadowRoot.appendChild(document.createElement("style")).textContent = ` 
```

Comment:
also here, use adopted stylesheet?

Reply:
you could use a helper like this: https://github.com/node-projects/base-custom-webcomponent/blob/master/src/BaseCustomWebComponent.ts#L28

then you could write the style in a nice way:

              style = css`
              :host {
              ....
              }`

Reply:
andy you could create it outside of the constructor and reuse it